### PR TITLE
ext_jwt: switch to new authlib

### PR DESCRIPTION
**Pull request by charandas**
**What is this feature?**

I started with a minimal changeset just to make my enterprise build pass. But seeing all the progress with namespaces on id tokens, I couldn't help but convert all the code to use the structs in authlib.

**Why do we need this feature?**

Reduces code sprawl.

**Who is this feature for?**

Identity Team

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -18,6 +18,9 @@ replace github.com/prometheus/prometheus => github.com/prometheus/prometheus v0.
 // Update when github.com/deepmap/oapi-codegen/v2 is updated to support later versions.
 replace github.com/getkin/kin-openapi => github.com/getkin/kin-openapi v0.122.0
 
+// TODO: remove after merge of https://github.com/grafana/authlib/pull/23
+replace github.com/grafana/authlib => /Users/charandas/go/src/github.com/grafana/authlib
```
Comment (charandas) [inline go.mod:22]: Had to add `Namespace` to `AccessTokenClaims` or else would have used the branch.

**Review by Jguer — APPROVED** [review]
Code seems ok on both sides. Approving for when the go.mod can be properly updated

**Review by gamab — APPROVED** [review]
LGTM, same as @Jguer  :slightly_smiling_face:

File: pkg/services/authn/clients/ext_jwt.go [hunk 2]
```diff
@@ -34,28 +34,39 @@ const (
...
-	})
+	}, authlib.TokenTypeAccess)
+
+	// The generic parameter is likely incorrect for ID Tokens
+	idTokenVerifier := authlib.NewVerifier[ExtendedJWTClaims](authlib.VerifierConfig{
+		SigningKeysURL: cfg.ExtJWTAuth.JWKSUrl,
+		AllowedAudiences: []string{
+			cfg.ExtJWTAuth.ExpectAudience,
+		},
+	}, authlib.TokenTypeID)
```
Comment (kalleep) [inline pkg/services/authn/clients/ext_jwt.go:50]: We should probably used the typed version for these. So the first one can be replaced by https://github.com/grafana/authlib/blob/main/authn/verifier_access_token.go and the second one by https://github.com/grafana/authlib/blob/main/authn/verifier_id_token.go.

It not a blocker for this pr. I can fix that in another pr :)

**Comment by charandas** [comment]
Also, I think a lot of these tests may need to be moved to authlib.

File: pkg/services/authn/clients/ext_jwt.go (hunk 2 above)
Comment (charandas) [reply pkg/services/authn/clients/ext_jwt.go]: I addressed it, I think. The only difference from your recommendation is that I used the interface with generics over concrete types to allow for mocking in tests.

File: pkg/services/authn/clients/ext_jwt.go [hunk 3]
```diff
@@ -95,8 +90,13 @@ func (s *ExtendedJWT) IsEnabled() bool {
 	return s.cfg.ExtJWTAuth.Enabled
 }
 
-func (s *ExtendedJWT) authenticateAsUser(idTokenClaims,
-	accessTokenClaims *ExtendedJWTClaims) (*authn.Identity, error) {
+func (s *ExtendedJWT) authenticateAsUser(idTokenClaims *authlib.Claims[authlib.IDTokenClaims],
+	accessTokenClaims *authlib.Claims[authlib.AccessTokenClaims]) (*authn.Identity, error) {
+	// Only allow token pairs with matching namespace through
+	if accessTokenClaims.Rest.Namespace != idTokenClaims.Rest.Namespace {
```
Comment (charandas) [inline pkg/services/authn/clients/ext_jwt.go:96]: Does this logic make sense to y'all? It shouldn't happen in practice but this is my failsafe to make sure that services don't send any all powerful tokens here (with `namespace = *`, that is).

File: pkg/services/authn/clients/ext_jwt.go (hunk 2 above)
Comment (kalleep) [reply pkg/services/authn/clients/ext_jwt.go]: Yep that is how we want to do it :100:

File: pkg/services/authn/clients/ext_jwt.go (hunk 3 above)
Comment (gamab) [reply pkg/services/authn/clients/ext_jwt.go:96]: I think we can start with that and loosen the check if a valid use case comes along.

**Pull request merged**